File: experiments/engine-kt/goal/src/main/kotlin/com/garive/eng/kt/goal/GoalLifecycle.kt
package com.garive.eng.kt.goal

/** Closed evidence family matching the four G1 criterion variants. */
public enum class GoalEvidenceKind(public val wireName: String) {
    /** Schema-validated user response evidence. */
    USER_ACCEPTANCE("user_acceptance"),
    /** Durable Artifact evidence. */
    ARTIFACT("artifact"),
    /** Durable Ledger fact evidence. */
    DURABLE_FACT("durable_fact"),
    /** Verified child-Goal terminal evidence. */
    CHILD_GOALS("child_goals"),
}

/** Exact evidence reference evaluated at a frozen commit version. */
public class GoalEvidenceV1 private constructor(
    public val evidenceId: GoalEvidenceId,
    public val criterionId: GoalCriterionId,
    public val kind: GoalEvidenceKind,
    public val durableReference: String,
    public val evidenceDigest: String,
    public val observedAtCommitVersion: Long,
) {
    public companion object {
        /** Validates a non-empty reference, canonical digest, and positive position. */
        public fun create(
            evidenceId: GoalEvidenceId,
            criterionId: GoalCriterionId,
            kind: GoalEvidenceKind,
            durableReference: String,
            evidenceDigest: String,
            observedAtCommitVersion: Long,
        ): GoalResult<GoalEvidenceV1> =
            if (durableReference.isEmpty() || !validDigest(evidenceDigest) || observedAtCommitVersion <= 0) {
                failure(GoalErrorCode.GOAL_INVALID)
            } else {
                GoalResult.Success(
                    GoalEvidenceV1(
                        evidenceId,
                        criterionId,
                        kind,
                        durableReference,
                        evidenceDigest,
                        observedAtCommitVersion,
                    ),
                )
            }
    }
}

/** Closed durable Goal lifecycle state. */
public enum class GoalState(public val wireName: String) {
    /** Definition may still be revised before work starts. */
    DRAFT("draft"),
    /** Work is currently admitted. */
    ACTIVE("active"),
    /** Typed external input or reconciliation is required. */
    SUSPENDED("suspended"),
    /** Every success criterion has verified evidence. */
    SUCCEEDED("succeeded"),
    /** Work ended unsuccessfully. */
    FAILED("failed"),
    /** Authenticated actor cancelled the Goal. */
    CANCELLED("cancelled"),
}

/** One requested pure lifecycle transition after Runtime command validation. */
public sealed interface GoalTransition {
    /** Start or resume work. */
    public data object Activate : GoalTransition
    /** Pause for one stable reason code. */
    public data class Suspend(public val reason: String) : GoalTransition
    /** Close successfully with complete exact evidence. */
    public data class Succeed(public val evidence: List<GoalEvidenceV1>) : GoalTransition
    /** Close unsuccessfully with one stable reason code. */
    public data class Fail(public val reason: String) : GoalTransition
    /** Cancel with one stable reason code. */
    public data class Cancel(public val reason: String) : GoalTransition
    /** Replace definition content and return to Draft. */
    public data class Revise(public val definition: GoalDefinitionV1) : GoalTransition
}

/** Immutable Goal projection after one contiguous durable prefix. */
public class GoalSnapshot private constructor(
    public val definition: GoalDefinitionV1,
    public val revision: Long,
    public val state: GoalState,
    terminalEvidence: List<GoalEvidenceV1>,
) {
    /** Exact success evidence, empty before Succeeded. */
    public val terminalEvidence: List<GoalEvidenceV1> = terminalEvidence.toList()

    /** Applies one transition only at the caller's exact expected revision. */
    public fun apply(expectedRevision: Long, transition: GoalTransition): GoalResult<GoalSnapshot> {
        if (expectedRevision != revision) return failure(GoalErrorCode.GOAL_REVISION_CONFLICT)
        if (state in TERMINAL_STATES) return failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
        val nextRevision = revision + 1
        if (nextRevision <= revision) return failure(GoalErrorCode.GOAL_INVALID)
        return when (transition) {
            GoalTransition.Activate -> if (state == GoalState.DRAFT || state == GoalState.SUSPENDED) {
                GoalResult.Success(GoalSnapshot(definition, nextRevision, GoalState.ACTIVE, emptyList()))
            } else {
                failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
            }
            is GoalTransition.Suspend -> if (state == GoalState.ACTIVE && transition.reason.isNotEmpty()) {
                GoalResult.Success(GoalSnapshot(definition, nextRevision, GoalState.SUSPENDED, emptyList()))
            } else {
                failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
            }
            is GoalTransition.Succeed -> if (state != GoalState.ACTIVE) {
                failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
            } else if (!validEvidence(definition.criteria, transition.evidence)) {
                failure(GoalErrorCode.GOAL_EVIDENCE_INSUFFICIENT)
            } else {
                GoalResult.Success(
                    GoalSnapshot(definition, nextRevision, GoalState.SUCCEEDED, transition.evidence),
                )
            }
            is GoalTransition.Fail -> if (
                state in setOf(GoalState.ACTIVE, GoalState.SUSPENDED) && transition.reason.isNotEmpty()
            ) {
                GoalResult.Success(GoalSnapshot(definition, nextRevision, GoalState.FAILED, emptyList()))
            } else {
                failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
            }
            is GoalTransition.Cancel -> if (transition.reason.isNotEmpty()) {
                GoalResult.Success(GoalSnapshot(definition, nextRevision, GoalState.CANCELLED, emptyList()))
            } else {
                failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
            }
            is GoalTransition.Revise -> if (transition.definition.goalId == definition.goalId) {
                GoalResult.Success(GoalSnapshot(transition.definition, nextRevision, GoalState.DRAFT, emptyList()))
            } else {
                failure(GoalErrorCode.GOAL_TRANSITION_INVALID)
            }
        }
    }

    public companion object {
        private val TERMINAL_STATES: Set<GoalState> =
            setOf(GoalState.SUCCEEDED, GoalState.FAILED, GoalState.CANCELLED)

        /** Creates revision 1 in Draft from a validated definition. */
        public fun create(definition: GoalDefinitionV1): GoalSnapshot =
            GoalSnapshot(definition, 1, GoalState.DRAFT, emptyList())
    }
}

private fun validEvidence(criteria: List<GoalCriterion>, evidence: List<GoalEvidenceV1>): Boolean {
    val byCriterion = evidence.associateBy(GoalEvidenceV1::criterionId)
    return byCriterion.size == evidence.size && evidence.map(GoalEvidenceV1::evidenceId).distinct().size == evidence.size &&
        criteria.size == evidence.size && criteria.all { criterion ->
            byCriterion[criterion.criterionId]?.kind == criterionKind(criterion)
        }
}

private fun criterionKind(value: GoalCriterion): GoalEvidenceKind = when (value) {
    is GoalCriterion.UserAcceptance -> GoalEvidenceKind.USER_ACCEPTANCE
    is GoalCriterion.Artifact -> GoalEvidenceKind.ARTIFACT
    is GoalCriterion.DurableFact -> GoalEvidenceKind.DURABLE_FACT
    is GoalCriterion.ChildGoals -> GoalEvidenceKind.CHILD_GOALS
}

File: engine/goal/tests/goal_lifecycle.rs
use garive_goal::{
    GoalBoundsV1, GoalCapabilityReference, GoalCriterion, GoalCriterionId, GoalDefinitionV1,
    GoalErrorCode, GoalEvidenceId, GoalEvidenceKind, GoalEvidenceV1, GoalId, GoalScopeV1,
    GoalSnapshot, GoalState, GoalTransition,
};

#[test]
fn activation_suspension_and_resume_advance_contiguous_revisions() {
    let draft = GoalSnapshot::new(definition("goal-1", "Ship the slice"));
    let active = draft.apply(1, GoalTransition::Activate).unwrap();
    let suspended = active
        .apply(2, GoalTransition::Suspend("approval_required".into()))
        .unwrap();
    let resumed = suspended.apply(3, GoalTransition::Activate).unwrap();
    assert_eq!(
        (resumed.revision(), resumed.state()),
        (4, GoalState::Active)
    );
}

#[test]
fn stale_revision_and_invalid_edges_fail_without_mutation() {
    let draft = GoalSnapshot::new(definition("goal-1", "Ship the slice"));
    assert_eq!(
        draft.apply(2, GoalTransition::Activate).unwrap_err().code(),
        GoalErrorCode::GoalRevisionConflict
    );
    assert_eq!(
        draft
            .apply(1, GoalTransition::Succeed(vec![evidence()]))
            .unwrap_err()
            .code(),
        GoalErrorCode::GoalTransitionInvalid
    );
}

#[test]
fn success_requires_exact_kind_and_complete_evidence() {
    let active = GoalSnapshot::new(definition("goal-1", "Ship the slice"))
        .apply(1, GoalTransition::Activate)
        .unwrap();
    assert_eq!(
        active
            .apply(2, GoalTransition::Succeed(Vec::new()))
            .unwrap_err()
            .code(),
        GoalErrorCode::GoalEvidenceInsufficient
    );
    let wrong = GoalEvidenceV1::new(
        GoalEvidenceId::new("evidence-1").unwrap(),
        GoalCriterionId::new("accepted").unwrap(),
        GoalEvidenceKind::Artifact,
        "artifact-1",
        digest('b'),
        5,
    )
    .unwrap();
    assert_eq!(
        active
            .apply(2, GoalTransition::Succeed(vec![wrong]))
            .unwrap_err()
            .code(),
        GoalErrorCode::GoalEvidenceInsufficient
    );
}

#[test]
fn verified_success_is_terminal() {
    let succeeded = GoalSnapshot::new(definition("goal-1", "Ship the slice"))
        .apply(1, GoalTransition::Activate)
        .unwrap()
        .apply(2, GoalTransition::Succeed(vec![evidence()]))
        .unwrap();
    assert_eq!(
        (succeeded.revision(), succeeded.state()),
        (3, GoalState::Succeeded)
    );
    assert_eq!(
        succeeded
            .apply(3, GoalTransition::Cancel("changed_mind".into()))
            .unwrap_err()
            .code(),
        GoalErrorCode::GoalTransitionInvalid
    );
}

#[test]
fn revision_replaces_definition_but_not_goal_identity() {
    let active = GoalSnapshot::new(definition("goal-1", "First objective"))
        .apply(1, GoalTransition::Activate)
        .unwrap();
    let revised = active
        .apply(
            2,
            GoalTransition::Revise(Box::new(definition("goal-1", "Revised objective"))),
        )
        .unwrap();
    assert_eq!((revised.revision(), revised.state()), (3, GoalState::Draft));
    assert_eq!(
        revised
            .apply(
                3,
                GoalTransition::Revise(Box::new(definition("goal-2", "Wrong identity"))),
            )
            .unwrap_err()
            .code(),
        GoalErrorCode::GoalTransitionInvalid
    );
}

#[test]
fn definition_digest_is_canonical_and_objective_sensitive() {
    assert_eq!(
        definition("goal-1", "Ship the slice").digest().unwrap(),
        "fa3b251bc520d2d66080f27ab1827b1c36e89c71a59f740e8570a77c8b42fe76"
    );
    assert_ne!(
        definition("goal-1", "Ship the slice").digest().unwrap(),
        definition("goal-1", "Ship another slice").digest().unwrap()
    );
}

#[test]
fn canonical_definition_round_trips_and_noncanonical_input_is_rejected() {
    let original = definition("goal-1", "Ship the slice");
    let json = original.canonical_json().unwrap();
    let decoded = GoalDefinitionV1::from_canonical_json(&json).unwrap();
    assert_eq!(decoded, original);
    assert_eq!(
        GoalDefinitionV1::from_canonical_json(&format!(" {json}"))
            .unwrap_err()
            .code(),
        GoalErrorCode::GoalInvalid
    );
}

fn definition(goal_id: &str, objective: &str) -> GoalDefinitionV1 {
    GoalDefinitionV1::new(
        GoalId::new(goal_id).unwrap(),
        objective,
        vec![GoalCriterion::UserAcceptance {
            criterion_id: GoalCriterionId::new("accepted").unwrap(),
            response_schema_digest: digest('a'),
        }],
        GoalScopeV1::new(Some("session-1".into()), ["workspace-1".into()]).unwrap(),
        GoalBoundsV1::new(3, 4, 2, Some(10_000), Some(60_000)).unwrap(),
        None,
        [GoalCapabilityReference::new("tools", "catalogue-v1").unwrap()],
    )
    .unwrap()
}

fn evidence() -> GoalEvidenceV1 {
    GoalEvidenceV1::new(
        GoalEvidenceId::new("evidence-1").unwrap(),
        GoalCriterionId::new("accepted").unwrap(),
        GoalEvidenceKind::UserAcceptance,
        "interaction-1",
        digest('b'),
        5,
    )
    .unwrap()
}

fn digest(character: char) -> String {
    character.to_string().repeat(64)
}
